File: Src/java/cql-to-elm/src/main/java/org/cqframework/cql/cql2elm/model/OperatorEntry.java
package org.cqframework.cql.cql2elm.model;

import java.util.*;
import org.hl7.cql.model.ChoiceType;
import org.hl7.cql.model.DataType;

public class OperatorEntry {
    public OperatorEntry(String name) {
        if (name == null || name.equals("")) {
            throw new IllegalArgumentException("name is null or empty");
        }

        this.name = name;
    }

    private String name;

    public String getName() {
        return this.name;
    }

    private SignatureNodes signatures = new SignatureNodes();
    private Map<Signature, GenericOperator> genericOperators = new HashMap<>();

    private static class SignatureNode {
        public SignatureNode(Operator operator) {
            if (operator == null) {
                throw new IllegalArgumentException("operator is null.");
            }

            this.operator = operator;
        }

        private Operator operator;

        public Operator getOperator() {
            return operator;
        }

        public Signature getSignature() {
            return operator.getSignature();
        }

        /*
        The invocation signature is the call signature with arguments of type Any set to the operand types
         */
        private Signature getInvocationSignature(Signature callSignature, Signature operatorSignature) {
            if (callSignature.getSize() == operatorSignature.getSize()) {
                DataType[] invocationTypes = new DataType[callSignature.getSize()];
                Iterator<DataType> callTypes = callSignature.getOperandTypes().iterator();
                Iterator<DataType> operatorTypes =
                        operatorSignature.getOperandTypes().iterator();
                boolean isResolved = false;
                for (int i = 0; i < invocationTypes.length; i++) {
                    DataType callType = callTypes.next();
                    DataType operatorType = operatorTypes.next();
                    if (callType.equals(DataType.ANY) && !operatorType.equals(DataType.ANY)) {
                        isResolved = true;
                        invocationTypes[i] = operatorType;
                    } else {
                        invocationTypes[i] = callType;
                    }
                }
                if (isResolved) {
                    return new Signature(invocationTypes);
                }
            }
            return callSignature;
        }

        private OperatorResolution getOperatorResolution(
                Operator operator,
                Signature callSignature,
                Signature invocationSignature,
                ConversionMap conversionMap,
                OperatorMap operatorMap,
                boolean allowPromotionAndDemotion,
                boolean requireConversions) {
            Conversion[] conversions = getConversions(
                    callSignature, operator.getSignature(), conversionMap, operatorMap, allowPromotionAndDemotion);
            OperatorResolution result = new OperatorResolution(operator, conversions);
            if (requireConversions && conversions == null) {
                return null;
            }
            return result;
        }

        public List<OperatorResolution> resolve(
                CallContext callContext, ConversionMap conversionMap, OperatorMap operatorMap) {
            List<OperatorResolution> results = null;
            Signature invocationSignature = getInvocationSignature(callContext.getSignature(), operator.getSignature());

            // Attempt exact match against this signature
            if (operator.getSignature().equals(invocationSignature)) {
                OperatorResolution result = getOperatorResolution(
                        operator,
                        callContext.getSignature(),
                        invocationSignature,
                        conversionMap,
                        operatorMap,
                        callContext.isAllowPromotionAndDemotion(),
                        false);
                if (result != null) {
                    results = new ArrayList<>();
                    results.add(result);
                    return results;
                }
            }

            // Attempt to resolve against sub signatures
            results = subSignatures.resolve(callContext, conversionMap, operatorMap);

            // If no subsignatures match, attempt subType match against this signature
            if (results == null && operator.getSignature().isSuperTypeOf(invocationSignature)) {
                OperatorResolution result = getOperatorResolution(
                        operator,
                        callContext.getSignature(),
                        invocationSignature,
                        conversionMap,
                        operatorMap,
                        callContext.isAllowPromotionAndDemotion(),
                        false);
                if (result != null) {
                    results = new ArrayList<>();
                    results.add(result);
                    return results;
                }
            }

            if (results == null && conversionMap != null) {
                // Attempt to find a conversion path from the call signature to the target signature
                OperatorResolution result = getOperatorResolution(
                        operator,
                        callContext.getSignature(),
                        invocationSignature,
                        conversionMap,
                        operatorMap,
                        callContext.isAllowPromotionAndDemotion(),
                        true);
                if (result != null) {
                    if (results == null) {
                        results = new ArrayList<>();
                    }
                    results.add(result);
                }
            }

            return results;
        }

        private Conversion[] getConversions(
                Signature callSignature,
                Signature operatorSignature,
                ConversionMap conversionMap,
                OperatorMap operatorMap,
                boolean allowPromotionAndDemotion) {
            if (callSignature == null
                    || operatorSignature == null
                    || callSignature.getSize() != operatorSignature.getSize()) {
                return null;
            }

            Conversion[] conversions = new Conversion[callSignature.getSize()];
            boolean isConvertible = callSignature.isConvertibleTo(
                    operatorSignature, conversionMap, operatorMap, allowPromotionAndDemotion, conversions);

            if (isConvertible) {
                return conversions;
            }

            return null;
        }

        private SignatureNodes subSignatures = new SignatureNodes();

        public boolean hasSubSignatures() {
            return subSignatures.hasSignatures();
        }

        @Override
        public int hashCode() {
            return operator.getSignature().hashCode();
        }

        @Override
        public boolean equals(Object o) {
            if (o instanceof SignatureNode) {
                SignatureNode that = (SignatureNode) o;
                return this.operator.getName().equals(that.operator.getName())
                        && this.getSignature().equals(that.getSignature());
            }

            return false;
        }

        public String toString() {
            return operator.toString();
        }
    }

    private static class SignatureNodes {
        private Map<Signature, SignatureNode> signatures = new HashMap<>();

        public boolean hasSignatures() {
            return signatures.size() > 0;
        }

        public boolean contains(Operator operator) {
            boolean result = signatures.containsKey(operator.getSignature());
            if (!result) {
                for (SignatureNode n : signatures.values()) {
                    result = n.subSignatures.contains(operator);
                    if (result) {
                        break;
                    }
                }
            }

            return result;
        }

        public void add(SignatureNode node) {
            if (node == null) {
                throw new IllegalArgumentException("node is null.");
            }

            if (signatures.containsKey(node.getSignature())) {
                throw new IllegalArgumentException(String.format(
                        "Operator %s already has a registration for signature: %s.",
                        node.operator.getName(), node.getSignature().toString()));
            }

            boolean added = false;
            for (SignatureNode n : signatures.values()) {
                if (n.getSignature().isSuperTypeOf(node.getSignature())) {
                    n.subSignatures.add(node);
                    added = true;
                    break;
                }
            }

            if (!added) {
                for (SignatureNode n : signatures.values().toArray(new SignatureNode[signatures.size()])) {
                    if (node.getSignature().isSuperTypeOf(n.getSignature())) {
                        signatures.remove(n.getSignature());
                        node.subSignatures.add(n);
                    }
                }

                signatures.put(node.getSignature(), node);
            }
        }

        public List<OperatorResolution> resolve(
                CallContext callContext, ConversionMap conversionMap, OperatorMap operatorMap) {
            ArrayList<OperatorResolution> results = null;

            int signatureCount = 0;
            for (SignatureNode n : signatures.values()) {

                if (n.getSignature().getSize() == callContext.getSignature().getSize()) {
                    signatureCount++;

                    // Any subSignature will count as an overload
                    if (n.hasSubSignatures()) {
                        signatureCount++;
                    }
                }

                List<OperatorResolution> nodeResults = n.resolve(callContext, conversionMap, operatorMap);
                if (nodeResults != null) {
                    if (results == null) {
                        results = new ArrayList<>();
                    }
                    results.addAll(nodeResults);
                }
            }

            if (results != null && signatureCount > 1) {
                for (OperatorResolution result : results) {
                    result.setOperatorHasOverloads();
                }
            }

            return results;
        }
    }

    public boolean containsOperator(Operator operator) {
        if (operator instanceof GenericOperator) {
            return containsGenericOperator((GenericOperator) operator);
        } else {
            return signatures.contains(operator);
        }
    }

    public void addOperator(Operator operator) {
        if (operator instanceof GenericOperator) {
            addGenericOperator((GenericOperator) operator);
        } else {
            signatures.add(new SignatureNode(operator));
        }
    }

    private boolean containsGenericOperator(GenericOperator operator) {
        return genericOperators.containsKey(operator.getSignature());
    }

    private void addGenericOperator(GenericOperator operator) {
        if (genericOperators.containsKey(operator.getSignature())) {
            throw new IllegalArgumentException(String.format(
                    "Operator %s already has a generic registration for signature: %s.",
                    name, operator.getSignature().toString()));
        }

        genericOperators.put(operator.getSignature(), operator);
    }

    private boolean allResultsUseConversion(List<OperatorResolution> results) {
        for (OperatorResolution resolution : results) {
            if (!resolution.hasConversions()) {
                return false;
            }
        }

        return true;
    }

    public List<Signature> expandChoices(Signature callSignature) {
        ArrayList<Signature> signatures = new ArrayList<Signature>();
        if (callSignature.containsChoices()) {

            ArrayList<ArrayList<DataType>> operandList = new ArrayList<ArrayList<DataType>>();
            for (DataType operand : callSignature.getOperandTypes()) {
                ArrayList<DataType> list = new ArrayList<DataType>();
                if (operand instanceof ChoiceType) {
                    for (DataType type : ((ChoiceType) operand).getTypes()) {
                        list.add(type);
                    }
                } else {
                    list.add(operand);
                }
                operandList.add(list);
            }

            DataType[] result = new DataType[callSignature.getSize()];
            collectSignatures(operandList, result, 0, signatures);
        } else {
            signatures.add(callSignature);
        }
        return signatures;
    }

    private void collectSignatures(
            ArrayList<ArrayList<DataType>> operandList, DataType[] result, int k, List<Signature> signatures) {
        if (k == operandList.size()) {
            signatures.add(new Signature(result));
        } else {
            for (int j = 0; j < operandList.get(k).size(); j++) {
                result[k] = operandList.get(k).get(j);
                collectSignatures(operandList, result, k + 1, signatures);
            }
        }
    }

    public List<OperatorResolution> resolve(
            CallContext callContext, OperatorMap operatorMap, ConversionMap conversionMap) {
        if (callContext == null) {
            throw new IllegalArgumentException("callContext is null");
        }

        // Attempt to instantiate any generic signatures
        // If the callContext signature contains choices, attempt instantiation with all possible combinations of
        // the call signature (ouch, this could really hurt...)
        boolean signaturesInstantiated = false;
        List<Signature> callSignatures = expandChoices(callContext.getSignature());
        for (Signature callSignature : callSignatures) {
            List<Operator> instantiations =
                    instantiate(callSignature, operatorMap, conversionMap, callContext.isAllowPromotionAndDemotion());
            for (Operator instantiation : instantiations) {
                // If the generic signature was instantiated, store it as an actual signature.
                if (!signatures.contains(instantiation)) {
                    signatures.add(new SignatureNode(instantiation));
                }
            }
        }

        return signatures.resolve(callContext, conversionMap, operatorMap);
    }

    private List<Operator> instantiate(
            Signature signature,
            OperatorMap operatorMap,
            ConversionMap conversionMap,
            boolean allowPromotionAndDemotion) {
        List<Operator> instantiations = new ArrayList<Operator>();

        for (GenericOperator genericOperator : genericOperators.values()) {
            InstantiationResult instantiationResult =
                    genericOperator.instantiate(signature, operatorMap, conversionMap, allowPromotionAndDemotion);
            if (instantiationResult.getOperator() != null) {
                instantiations.add(instantiationResult.getOperator());
            }
        }

        return instantiations;
    }
}

File: cql-to-elm/src/commonMain/kotlin/org/cqframework/cql/cql2elm/model/OperatorEntry.kt
package org.cqframework.cql.cql2elm.model

import kotlin.collections.ArrayList
import kotlin.collections.HashMap
import org.hl7.cql.model.ChoiceType
import org.hl7.cql.model.DataType

class OperatorEntry(val name: String) {

    private val signatures = SignatureNodes()
    private val genericOperators: MutableMap<Signature, GenericOperator> = HashMap()

    init {
        require(name.isNotEmpty()) { "name is empty" }
    }

    private class SignatureNode(val operator: Operator) {
        val signature: Signature
            get() = operator.signature

        /*
        The invocation signature is the call signature with arguments of type Any set to the operand types
         */
        private fun getInvocationSignature(
            callSignature: Signature,
            operatorSignature: Signature,
        ): Signature {
            if (callSignature.size == operatorSignature.size) {
                val invocationTypes = mutableListOf<DataType>()
                val callTypes = callSignature.operandTypes
                val operatorTypes = operatorSignature.operandTypes
                var isResolved = false
                for (i in callTypes.indices) {
                    val callType = callTypes[i]
                    val operatorType = operatorTypes[i]
                    if (callType == DataType.ANY && operatorType != DataType.ANY) {
                        isResolved = true
                        invocationTypes.add(operatorType)
                    } else {
                        invocationTypes.add(callType)
                    }
                }
                if (isResolved) {
                    return Signature(invocationTypes)
                }
            }
            return callSignature
        }

        @Suppress("LongParameterList", "UnusedParameter")
        private fun getOperatorResolution(
            operator: Operator,
            callSignature: Signature,
            invocationSignature: Signature,
            conversionMap: ConversionMap,
            operatorMap: OperatorMap,
            allowPromotionAndDemotion: Boolean,
            requireConversions: Boolean,
        ): OperatorResolution? {
            val conversions =
                getConversions(
                    callSignature,
                    operator.signature,
                    conversionMap,
                    operatorMap,
                    allowPromotionAndDemotion,
                )
            if (requireConversions && conversions == null) {
                return null
            }

            val result = OperatorResolution(operator, conversions)
            return result
        }

        @Suppress("ReturnCount")
        fun resolve(
            callContext: CallContext,
            conversionMap: ConversionMap,
            operatorMap: OperatorMap,
        ): List<OperatorResolution> {
            var results: MutableList<OperatorResolution> = ArrayList()
            val invocationSignature =
                getInvocationSignature(callContext.signature, operator.signature)

            // Attempt exact match against this signature
            if (operator.signature == invocationSignature) {
                val result =
                    getOperatorResolution(
                        operator,
                        callContext.signature,
                        invocationSignature,
                        conversionMap,
                        operatorMap,
                        callContext.allowPromotionAndDemotion,
                        false,
                    )
                if (result != null) {
                    results.add(result)
                    return results
                }
            }

            // Attempt to resolve against sub signatures
            results = subSignatures.resolve(callContext, conversionMap, operatorMap)

            // If no subsignatures match, attempt subType match against this signature
            if (results.isEmpty() && operator.signature.isSuperTypeOf(invocationSignature)) {
                val result =
                    getOperatorResolution(
                        operator,
                        callContext.signature,
                        invocationSignature,
                        conversionMap,
                        operatorMap,
                        callContext.allowPromotionAndDemotion,
                        false,
                    )
                if (result != null) {
                    results.add(result)
                    return results
                }
            }

            if (results.isEmpty()) {
                // Attempt to find a conversion path from the call signature to the target signature
                val result =
                    getOperatorResolution(
                        operator,
                        callContext.signature,
                        invocationSignature,
                        conversionMap,
                        operatorMap,
                        callContext.allowPromotionAndDemotion,
                        true,
                    )
                if (result != null) {
                    results.add(result)
                }
            }

            return results
        }

        @Suppress("ReturnCount")
        private fun getConversions(
            callSignature: Signature?,
            operatorSignature: Signature?,
            conversionMap: ConversionMap,
            operatorMap: OperatorMap,
            allowPromotionAndDemotion: Boolean,
        ): Array<Conversion?>? {
            if (
                callSignature == null ||
                    operatorSignature == null ||
                    callSignature.size != operatorSignature.size
            ) {
                return null
            }

            val conversions = arrayOfNulls<Conversion>(callSignature.size)
            val isConvertible =
                callSignature.isConvertibleTo(
                    operatorSignature,
                    conversionMap,
                    operatorMap,
                    allowPromotionAndDemotion,
                    conversions,
                )

            if (isConvertible) {
                return conversions
            }

            return null
        }

        val subSignatures: SignatureNodes = SignatureNodes()

        fun hasSubSignatures(): Boolean {
            return subSignatures.hasSignatures()
        }

        override fun hashCode(): Int {
            return operator.signature.hashCode()
        }

        override fun equals(other: Any?): Boolean {
            if (other is SignatureNode) {
                return operator.name == other.operator.name && (this.signature == other.signature)
            }

            return false
        }

        override fun toString(): String {
            return operator.toString()
        }
    }

    private class SignatureNodes {
        private val signatures: MutableMap<Signature, SignatureNode> = HashMap()

        fun hasSignatures(): Boolean {
            return signatures.isNotEmpty()
        }

        fun contains(operator: Operator?): Boolean {
            var result = signatures.containsKey(operator!!.signature)
            if (!result) {
                for (n in signatures.values) {
                    result = n.subSignatures.contains(operator)
                    if (result) {
                        break
                    }
                }
            }

            return result
        }

        fun add(node: SignatureNode?) {
            requireNotNull(node) { "node is null." }

            require(!signatures.containsKey(node.signature)) {
                "Operator ${node.operator.name} already has a registration for signature: ${node.signature}."
            }

            var added = false
            for (n in signatures.values) {
                if (n.signature.isSuperTypeOf(node.signature)) {
                    n.subSignatures.add(node)
                    added = true
                    break
                }
            }

            if (!added) {
                for (n in signatures.values.toTypedArray<SignatureNode>()) {
                    if (node.signature.isSuperTypeOf(n.signature)) {
                        signatures.remove(n.signature)
                        node.subSignatures.add(n)
                    }
                }

                signatures[node.signature] = node
            }
        }

        fun resolve(
            callContext: CallContext,
            conversionMap: ConversionMap,
            operatorMap: OperatorMap,
        ): MutableList<OperatorResolution> {
            val results = arrayListOf<OperatorResolution>()

            var signatureCount = 0
            for (n in signatures.values) {
                if (n.signature.size == callContext.signature.size) {
                    signatureCount++

                    // Any subSignature will count as an overload
                    if (n.hasSubSignatures()) {
                        signatureCount++
                    }
                }

                val nodeResults = n.resolve(callContext, conversionMap, operatorMap)
                results.addAll(nodeResults)
            }

            if (signatureCount > 1) {
                for (result in results) {
                    result.operatorHasOverloads = true
                }
            }

            return results
        }
    }

    fun containsOperator(operator: Operator?): Boolean {
        return if (operator is GenericOperator) {
            containsGenericOperator(operator)
        } else {
            signatures.contains(operator)
        }
    }

    fun addOperator(operator: Operator) {
        if (operator is GenericOperator) {
            addGenericOperator(operator)
        } else {
            signatures.add(SignatureNode(operator))
        }
    }

    private fun containsGenericOperator(operator: GenericOperator): Boolean {
        return genericOperators.containsKey(operator.signature)
    }

    private fun addGenericOperator(operator: GenericOperator) {
        require(!genericOperators.containsKey(operator.signature)) {
            "Operator $name already has a generic registration for signature: ${operator.signature}."
        }

        genericOperators[operator.signature] = operator
    }

    @Suppress("NestedBlockDepth")
    private fun expandChoices(callSignature: Signature): List<Signature> {
        val signatures = ArrayList<Signature>()
        if (callSignature.containsChoices) {
            val operandList = ArrayList<ArrayList<DataType>>()
            for (operand in callSignature.operandTypes) {
                val list = ArrayList<DataType>()
                if (operand is ChoiceType) {
                    for (type in operand.types) {
                        list.add(type)
                    }
                } else {
                    list.add(operand)
                }
                operandList.add(list)
            }

            val result = arrayOfNulls<DataType>(callSignature.size)
            collectSignatures(operandList, result, 0, signatures)
        } else {
            signatures.add(callSignature)
        }
        return signatures
    }

    private fun collectSignatures(
        operandList: ArrayList<ArrayList<DataType>>,
        result: Array<DataType?>,
        k: Int,
        signatures: MutableList<Signature>,
    ) {
        if (k == operandList.size) {
            val noNulls = result.toList().requireNoNulls()
            signatures.add(Signature(noNulls))
        } else {
            for (j in operandList[k].indices) {
                result[k] = operandList[k][j]
                collectSignatures(operandList, result, k + 1, signatures)
            }
        }
    }

    fun resolve(
        callContext: CallContext,
        operatorMap: OperatorMap,
        conversionMap: ConversionMap,
    ): List<OperatorResolution> {
        // Attempt to instantiate any generic signatures
        // If the callContext signature contains choices, attempt instantiation with all possible
        // combinations of
        // the call signature (ouch, this could really hurt...)
        val callSignatures = expandChoices(callContext.signature)
        for (callSignature in callSignatures) {
            val instantiations =
                instantiate(
                    callSignature,
                    operatorMap,
                    conversionMap,
                    callContext.allowPromotionAndDemotion,
                )
            for (instantiation in instantiations) {
                // If the generic signature was instantiated, store it as an actual signature.
                if (!signatures.contains(instantiation)) {
                    signatures.add(SignatureNode(instantiation))
                }
            }
        }

        return signatures.resolve(callContext, conversionMap, operatorMap)
    }

    private fun instantiate(
        signature: Signature,
        operatorMap: OperatorMap,
        conversionMap: ConversionMap,
        allowPromotionAndDemotion: Boolean,
    ): List<Operator> {
        return genericOperators.values
            .map {
                it.instantiate(signature, operatorMap, conversionMap, allowPromotionAndDemotion)
            }
            .mapNotNull { it.operator }
    }
}
